[WINSPOOL] Fix existing Level checks

SergeGautherie opened the pull request:
Based on MSDN.

Please, double-check `EnumPrintersW` and `GetPrinterDriverW`, where old `ASSERT` did not match.

SergeGautherie commented:
I added 2*2 `FIXME`s to `.idl`. Should I simply add the missing structures?

ThFabba commented on sdk/include/reactos/idl/winspool.idl:
```diff
@@ -74,6 +74,8 @@ typedef struct _WINSPOOL_DOC_INFO_1 {
 }
 WINSPOOL_DOC_INFO_1;
 
+/* FIXME: Why are WINSPOOL_DOC_INFO_2 and WINSPOOL_DOC_INFO_3 missing? */
```
These don't need to start a converstaion. Just the struct name, maybe with the word "missing" is enough.
Unless you're not sure that they're missing, in which case, be sure or don't bother adding comments.

SergeGautherie replied:
> Just the struct name

Done. Added one more case.

ThFabba commented on win32ss/printing/base/winspool/printers.c:
```diff
@@ -424,10 +424,9 @@ EnumPrintersA(DWORD Flags, PSTR Name, DWORD Level, PBYTE pPrinterEnum, DWORD cbB
 
     // Check for invalid levels here for early error return. MSDN says that only 1, 2, 4, and 5 are allowable.
-    if (Level !=  1 && Level != 2 && Level != 4 && Level != 5)
+    if (Level == 0 || Level == 3 || Level > 5)
```
Do I need to point out what's wrong here?

SergeGautherie replied:
> Do I need to point out what's wrong here?

Reverted, though I think consistency would be worth it here.

ThFabba commented on win32ss/printing/base/winspool/printers.c:
```diff
@@ -874,7 +873,6 @@ EnumPrintersW(DWORD Flags, PWSTR Name, DWORD Level, PBYTE pPrinterEnum, DWORD cb
     if (dwErrorCode == ERROR_SUCCESS)
     {
         // Replace relative offset addresses in the output by absolute pointers.
-        ASSERT(Level <= 9);
```
These seem to assert a condition about each specific `MarshallUpStructuresArray` call, so it may make sense to keep them.

SergeGautherie replied:
> assert

Right, reverted. Made them consistent with the related 'Marshalling' arrays, not local `Level`s.

SergeGautherie commented:
> Just the struct name

Done. Added one more case.

> Do I need to point out what's wrong here?

Reverted, though I think consistency would be worth it here.

> assert

Right, reverted. Made them consistent with the related 'Marshalling' arrays, not local `Level`s.

ColinFinck commented on win32ss/printing/include/marshalling/printerdrivers.h:
```diff
@@ -70,6 +70,7 @@ static const MARSHALLING PrinterDriver5Marshalling = {
     }
 };
 
+/* TODO: Add PrinterDriver6Marshalling and PrinterDriver8Marshalling */
```
I'm fine with accepting TODOs for stuff that is supported in NT5.x.
However, `DRIVER_INFO_8` was only added in NT6 according to MSDN.

On the other hand, there are structures like `JOB_INFO_4`, which also exist in NT5.x, but don't exist in our printing implementation for whatever reason and also not in your TODO comments.
If you already add TODOs for unimplemented printing structures, please check that you haven't missed any NT5.x one.

ColinFinck commented on win32ss/printing/base/spoolsv/jobs.c:
```diff
@@ -59,7 +59,7 @@ _RpcEnumJobs(WINSPOOL_PRINTER_HANDLE hPrinter, DWORD FirstJob, DWORD NoJobs, DWO
         if (Level <= 2)
         {
-            ASSERT(Level >= 1);
+            ASSERT(Level >= 1 && Level <= 2);
```
Guess cppcheck would issue a warning here due to a redundant condition :D

ColinFinck commented on win32ss/printing/base/spoolsv/printerdrivers.c:
```diff
@@ -63,6 +63,7 @@ _RpcGetPrinterDriver(WINSPOOL_PRINTER_HANDLE hPrinter, WCHAR* pEnvironment, DWOR
     if (GetPrinterDriverW(hPrinter, pEnvironment, Level, pDriverAligned, cbBuf, pcbNeeded))
     {
         // Replace relative offset addresses in the output by absolute pointers.
+        // TODO: Support 6 and 8 too.
```
See below.

ColinFinck commented on win32ss/printing/base/winspool/jobs.c:
```diff
@@ -92,7 +98,7 @@ EnumJobsW(HANDLE hPrinter, DWORD FirstJob, DWORD NoJobs, DWORD Level, PBYTE pJob
         if (Level <= 2)
         {
-            ASSERT(Level >= 1);
+            ASSERT(Level >= 1 && Level <= 2);
```
See above

ThFabba commented:
Changes requested more than two weeks ago. Please reopen if/when comments are addressed.
